### DISCUSSION TITLE
How would i go about adding an shortcut to windows terminal?

### DISCUSSION THREAD

**Discussion by Raukie** [microsoft/terminal#16936] [Q&A] [answered]
Hello, 
i have been look at the code and issues for the past few hours. But i still don't know what the best way is for me to add some custom functionality. I would like to add an custom keybind to windows terminal. 

I wouldn't mind doing it as an installable package (If that is even possible), or forking and changing the source code directly.
My idea is to add an shorcut that opens an file explorer window. Here you can select a file and it will copy the path to the current cursor position in the terminal. 

This is mostly for educational purposes and getting to know some production C++.
Thank you in advance.

**Comment by zadjii-msft** [comment, marked as answer]
There are some thoughts that might be relevant in #12859.

But also, if I can try to dig in to what you're trying to accomplish here - You want to select some text, then try to open that selected text in file explorer? In case you didn't know, the `start` command in a shell will open the path provided in file explorer. So `start c:\users\foo\downloads` opens `c:\users\foo\downloads` in File Explorer. It might be easier to just select the text, <kbd>ctrl+c</kbd> to copy, then <kbd>home</kbd> and type `start `

You might even be able to whip up a `multipleActions` command that does the `copy`, then `sendInput` the sequence `"\u001b[Hstart "` (which is the enoding for the <kbd>home</kbd> key).

**Reply by Raukie** [reply]
Sorry for the confusion in my orignal question. 
My idea is an shortcut that opens the file explorer. Where you will have the ability to select an file, this will use no input from the command line whatsoever(Except for the keybind specified to initiate the file explorer dialog).

When you select an file in the fileexplorer dialog. The file explorer window will close and the absolute path of the selected file will be copied into the terminal at your current cursor position. 

I hope this makes it more clear what the general idea is. This was also the reason why i think the normal command system is not enough to create this. Since the command line has to copy the path of the selected file to the current cursor position.

**Reply by Raukie** [reply]
Also i just looked at the referenced issue and it looks be to be doing, what i thought i needed to do too. I have been ofcourse looking through the code. But i was a bit overwhelmed with the amount of places i have to specify the shortcut i want to add to make it show up and have a starting point for my code (Could not find documentation on this, could definitly be because of me). 

I have read this [Comment in issue](https://github.com/microsoft/terminal/issues/12859#issuecomment-1111023499) which states there are some macros for adding an command but those are outdated. This is however an comment from 2022. Are there any new macros added. Or has the documentation stated in this issue as slightly outdated been updated?

**Reply by Raukie** [reply]
I have been able to make progress by following your commits in this issue about adding an ExportBuffer: https://github.com/microsoft/terminal/pull/12097

**Comment by Raukie** [comment]
Hello, 
For anyone interested i have succes. I am not sure if people would like this feature but i personally would.
This is the working example of my idea implemented in my local fork. 

https://github.com/microsoft/terminal/assets/90243019/c4fce082-27f9-48d8-ac63-56a1ad5eb469

https://github.com/microsoft/terminal/commit/7d0ff37f643ea5d1ae89878cf35d1f8483086e5b

I had a lot of fun making this and would love to hear what you think of the feature.

**Reply by Raukie** [reply]
I have now also added an check if quotes are necessary or not. And code for WSL support. Is not working though since i can't find an accessibel function to check if it is wsl. There is one private _interactivity which i could use but i don't know the best practices around changing that to public. 

https://github.com/microsoft/terminal/commit/67cf0d0a5ed03ab1647de2ebb3434d2cb5c9c07c
